homeassistant/components/mailbox/__init__.py

```diff
@@ -0,0 +1,175 @@
+"""
+Provides functionality for mailboxes.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/mailbox/
+"""
+
+import asyncio
+import logging
+from functools import partial
+from contextlib import suppress
+from datetime import timedelta
+
+import async_timeout
+
+from aiohttp import web
+
+from homeassistant.const import (HTTP_BAD_REQUEST)
+
+from homeassistant.helpers import discovery
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.entity import Entity
+from homeassistant.components.http import HomeAssistantView
+
+
+DOMAIN = 'mailbox'
+CONTENT_TYPE_MPEG = 'audio/mpeg'
+SCAN_INTERVAL = timedelta(seconds=30)
+_LOGGER = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Track states and offer events for mailboxes."""
+    component = EntityComponent(
+        logging.getLogger(__name__), DOMAIN, hass, SCAN_INTERVAL)
+
+    hass.components.frontend.register_built_in_panel(
+        'mailbox', 'Mailbox', 'mdi:account-location')
+    hass.http.register_view(MailboxMessageView(component.entities))
+    hass.http.register_view(MailboxMediaView(component.entities))
+    hass.http.register_view(MailboxDeleteView(component.entities))
+
+    yield from component.async_setup(config)
+
+    return True
+
+
+class Mailbox(Entity):
+    """Represent an mailbox device."""
+
+    def __init__(self, hass, domain, config):
+        """Init the Mailbox data object."""
+        self._domain = domain
+        self._config = config
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        yield from discovery.async_load_platform(
+            self.hass, "sensor", DOMAIN,
+            {'domain': self._domain, 'mailbox_id': self.entity_id},
+            self._config)
+
+    @property
+    def state(self):
+        """Return the state of the binary sensor."""
+        return str(len(self.get_messages()))
+
+    @property
+    def hidden(self) -> bool:
+        """Return whether the component should be hidden."""
+        return True
```

Comment:
You will have to remove the mailbox sensor completely. Components are not allowed to have sensors in other platforms that forward other platform data. 

If you still want to have an entity to write the current unread messages to the state machine, then that's a responsibility of the mailbox component to create entities like `mailbox.platform_name` with as state the number of unread messages.

However, you cannot combine the entity and fetching of messages in 1 object. You can have the entity poll the service object for new messages and then update the state.